Fix running bootstrap tests with a local Rust toolchain as the stage0

emilyalbini opened the pull request:
When configuring a local Rust toolchain as the stage0 (with `build.rustc` and `build.cargo` in `config.toml`) we noticed there were test failures (both on the Python and the Rust side) due to bootstrap not being able to find rustc and Cargo.

This was due to those two `config.toml` settings not being propagated in the tests. This PR fixes the issue by ensuring rustc and cargo are always configured in tests, using the parent bootstrap's `initial_rustc` and `initial_cargo`.

try-job: x86_64-msvc
Fixes https://github.com/rust-lang/rust/issues/105766

emilyalbini committed:
fix python bootstrap tests requiring a downloaded stage0

emilyalbini committed:
fix rust bootstrap tests requiring a downloaded stage0

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9514812792/job/26227751596) [(plain)](https://github.com/rust-lang/rust/commit/4f30d2181288202e6b2ec686f9dd3a6aee49aad4/checks/26227751596/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.406   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.418 Collecting boolean-py==4.0
#13 0.423   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.432 Collecting chardet==5.1.0
---
#13 3.430 Building wheels for collected packages: reuse
#13 3.431   Building wheel for reuse (pyproject.toml): started
#13 3.758   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.759   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.759   Stored in directory: /tmp/pip-ephem-wheel-cache-bbplwwm3/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.761 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.783   Attempting uninstall: setuptools
#13 3.783     Found existing installation: setuptools 59.6.0
#13 3.785     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.785     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.785     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.462 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.462 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.957 Collecting virtualenv
#13 4.996   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.056      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 68.0 MB/s eta 0:00:00
#13 5.091 Collecting distlib<1,>=0.3.7
#13 5.095   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.107      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 47.2 MB/s eta 0:00:00
#13 5.135 Collecting platformdirs<5,>=3.9.1
#13 5.138   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.170 Collecting filelock<4,>=3.12.2
#13 5.174   Downloading filelock-3.15.1-py3-none-any.whl (15 kB)
#13 5.261 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.426 Successfully installed distlib-0.3.8 filelock-3.15.1 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      219072 kB
DirectMap2M:     7120896 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/d2ad293851dc8e14a61355d0358490b77efae8cb/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-d2ad293851dc8e14a61355d0358490b77efae8cb-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5069 files
tidy check
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
linting python files
/checkout/src/bootstrap/bootstrap_test.py:158:13: F821 Undefined name `configure__args`
/checkout/src/bootstrap/bootstrap_test.py:158:13: F841 Local variable `configure__args` is assigned to but never used


python linting failed! Printing diff suggestions:
tidy error: checks with external tool 'ruff' failed
some tidy checks failed
  local time: Fri Jun 14 10:46:55 UTC 2024
  network time: Fri, 14 Jun 2024 10:46:55 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:pushpin: Commit fcb6ff5171fd41c988a6c449f76ff34c2c2d5a43 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan fcb6ff5171fd41c988a6c449f76ff34c2c2d5a43 -->
<!-- homu: {"type":"Approved","sha":"fcb6ff5171fd41c988a6c449f76ff34c2c2d5a43","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

cuviper commented:
I suppose this will fix #105766?

matthiaskrgr commented:
@bors r- I guess
https://github.com/rust-lang/rust/pull/126486#issuecomment-2168952828

onur-ozkan commented:
Seems like it's missing `.exe` part.

emilyalbini committed:
set the correct executable for initial_{rustc,cargo}

Due to the way the paths initial_rustc and initial_cargo were
constructed before this commit, they mixed \ and / for path separators
and they omitted the .exe suffix.

This worked fine up until now, as Windows is capable of handling the
mixed path separators and the Command::new API adds the ".exe" suffix if
missing from the executable.

This resulted in paths that didn't actually exist on disk though, due to
the missing .exe suffix. This commit fixes that by adding the .exe
suffix to initial_rustc and initial_cargo when --build is Windows.

pietroalbini commented:
> I suppose this will fix https://github.com/rust-lang/rust/issues/105766?

Yes!

bors commented:
:hourglass: Trying commit 198c809dd1b06d398c44f570c9635af640895ce9 with merge b7d2a8118e1fc850a3d8b4bf60c7ca265fc7f428...
<!-- homu: {"type":"TryBuildStarted","head_sha":"198c809dd1b06d398c44f570c9635af640895ce9","merge_sha":"b7d2a8118e1fc850a3d8b4bf60c7ca265fc7f428"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9840480429/job/27169743086)
Build commit: b7d2a8118e1fc850a3d8b4bf60c7ca265fc7f428 (`b7d2a8118e1fc850a3d8b4bf60c7ca265fc7f428`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9840480429/job/27169743086"},"merge_sha":"b7d2a8118e1fc850a3d8b4bf60c7ca265fc7f428"} -->

pietroalbini commented:
@rustbot ready

pietroalbini commented:
@rustbot author

Figured out another case that'd need tweaking.

emilyalbini committed:
set the correct rustc and cargo even for tests invoking parse_inner

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9842434744/job/27171362659) [(plain)](https://github.com/rust-lang/rust/commit/aed27c59abd058c2ae8af96cc87d19235cc5d1df/checks/27171362659/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.50s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
```

</details>

pietroalbini commented:
Confirmed this fixes all of the problems with using custom rustc/cargo in an airgapped environment.

@rustbot review

onur-ozkan commented on src/bootstrap/src/core/config/config.rs:
```diff
@@ -1440,14 +1451,24 @@ impl Config {
             config.out = crate::utils::helpers::absolute(&config.out);
         }
 
+        // Hacky way to determine the executable suffix for the build target. We cannot use
+        // std::env::consts::EXE_SUFFIX as the build target might not be the target bootstrap was
+        // compiled with.
+        let initial_exe_suffix = if config.build.triple.contains("windows") { ".exe" } else { "" };
```
You can use `utils::helpers::exe` function.

ref:
https://github.com/rust-lang/rust/blob/b4b2643c111714b541db1c18c178ad0ce654238b/src/bootstrap/src/core/config/config.rs#L1941

pietroalbini replied:
TIL about that helper! Updated.

emilyalbini committed:
use utils::helpers::exe

onur-ozkan commented:
LGTM, thanks!

@bors r+

bors commented:
:pushpin: Commit 9cd1d253a60aca27086da57b68c24b0924277d58 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 9cd1d253a60aca27086da57b68c24b0924277d58 -->
<!-- homu: {"type":"Approved","sha":"9cd1d253a60aca27086da57b68c24b0924277d58","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
